[FLINK-12029][table] Add column operations for TableApi

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request adds column operations for TableApi. Column oprations can be used to select or deselect columns. More details can be found in the google design doc.


## Brief change log

  - Add different columns operations for TableApi.
  - Add Tests
  - Add document


## Verifying this change


This change added tests and can be verified as follows:

  - Added integration test `testColumnOperations` in `CalcITCase`.
  - Add `ColumnsOperationTest` to test column operations in different table operations. 
  - Add validation tests `ColumnsOperationValidationTest` to test exceptions.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

sunjincheng121 commented on docs/dev/table/tableApi.md:
```diff
@@ -1682,6 +1682,286 @@ The `OverWindow` defines a range of rows over which aggregates are computed. `Ov
+      </td>
+      <td>
+{% highlight scala %}
+select(columns(2 ~ 4)) = select('b, 'c, 'd)
```
Please pay an attention on the comments in [google doc ](https://docs.google.com/document/d/1tryl6swt1K1pw7yvv5pdvFXSxfrBZ3_OkOObymis2ck/edit#), Timo suggest tousing some real word instead `~`,  Like `to` or `until`, +1 from my side, what do you think?

sunjincheng121 commented:
@flinkbot approve description
@flinkbot approve consensus
Hi, @hequn8128 Thanks for the PR，before starting the detailed review, I suggest paying attention to the comments in the google doc, due to there is some detail discussing for this JIRA. (such as the comment I left above.) :)

Best,Jincheng

sunjincheng121 commented:
@flinkbot attention @twalthr

sunjincheng121 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/batch/table/CalcITCase.scala:
```diff
@@ -593,6 +593,27 @@ class CalcITCase(
   }
+
+  @Test
+  def testColumnOperations(): Unit = {
```
I think we can remove the ITCase, due to the column operation only the Table API level Features， do not touch runtime logic. What do you think?

sunjincheng121 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/table/ColumnsOperationTest.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+  @Test
+  def testOrderBy(): Unit = {
```
We can only keep OrderBy test, due to other operator can be test by stream/table/ColumnsOperationsTest, What do you think?

sunjincheng121 commented on docs/dev/table/tableApi.md:
```diff
@@ -1682,6 +1682,286 @@ The `OverWindow` defines a range of rows over which aggregates are computed. `Ov
+{% highlight scala %}
+table
+   .groupBy(columns(1 ~ 3))
+   .select(columns('a ~ 'b), myUDAgg(myUDF(columns(5 ~ 20))))
```
Using the same example with JAVA?

sunjincheng121 commented:
@flinkbot approve-until

hequn8128 commented:
@sunjincheng121 Thanks a lot for your review. I have updated the PR according to your suggestions. 

Best, Hequn

sunjincheng121 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ColumnsOperationExpander.java:
```diff
@@ -0,0 +1,173 @@
+ * Expands column expressions to it's real parent's input references.
+ */
+@Internal
+public class ColumnsOperationExpander extends ApiExpressionDefaultVisitor<List<Expression>> {
```
How about move `ColumnsOperationExpander` into `ApiExpressionUtils`?

dawidwys replied:
This should be another `ResolverRule` once #8062 is merged.

hequn8128 replied:
@sunjincheng121 @dawidwys Thanks a lot for your advice. Maybe we can keep `ColumnsOperationExpander` here for now and replace it to a ResolverRule once #8062 is merged?

sunjincheng121 replied:
Make sense for my side. That depends on which PR be merged first. another should rebase code.

dawidwys replied:
I would be strongly in favor merging this after #8062. This PR introduces a new complex feature. If we merge it before #8062 it immediately introduces technical debt.

dawidwys replied:
I think this whole class should be internal to the rule.

hequn8128 replied:
This class is same with `LookupCallResolver` which is used not only in Rule but also used in TableImpl to make sure `extractFieldReferences` get the right results.

hequn8128 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/BuiltInFunctionDefinitions.java:
```diff
@@ -319,6 +319,14 @@
+		new FunctionDefinition("columnsRange", OTHER_FUNCTION);
+	public static final FunctionDefinition COLUMNS_INVERSE_SELECTION =
+		new FunctionDefinition("columnsInverseSelection", OTHER_FUNCTION);
+
```
Using `MINUS_PREFIX` directly is better. Thanks.

hequn8128 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/BuiltInFunctionDefinitions.java:
```diff
@@ -319,6 +319,14 @@
+	// columns
+	public static final FunctionDefinition COLUMNS =
+		new FunctionDefinition("columns", OTHER_FUNCTION);
+	public static final FunctionDefinition COLUMNS_RANGE =
```
Maybe we can name it `RANG_TO`. We may add `RANGE_UNTIL` later. What do you think?

sunjincheng121 replied:
+1 for `RANG_TO`.

hequn8128 commented:
@sunjincheng121 Thanks for your valuable suggestions. It makes the PR much better. I have addressed all your comments and updated the PR. Would be great if you can take another look. Thank you.

sunjincheng121 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ColumnsOperationExpander.java:
```diff
@@ -0,0 +1,177 @@
+			int start = (int) ((ValueLiteralExpression) range.getChildren().get(0)).getValue();
+			int end = (int) ((ValueLiteralExpression) range.getChildren().get(1)).getValue();
+			Preconditions.checkArgument(
+				start < end,
```
`start < end` -> `start <= end`,

hequn8128 commented:
@sunjincheng121 Sounds good. I will rebase the PR once [FLINK-11884](https://github.com/apache/flink/pull/8062/files) is merged. Would be great if @dawidwys can also take a look at that time. 

Best, Hequn

dawidwys commented:
Thank you for understanding! Will definitely look once it will be rebased on the merged changes.

dawidwys commented:
Hi @hequn8128,
The commit(413e3feb786df9b0df951eae7e942ac45ba3d22a) that introduces `ExpressionResolver` and `ResolverRules` was merged. I think you can rebase those changes on top of current master now and proceed with this PR.

Also thank you again for waiting for that change!

hequn8128 commented:
@dawidwys Cool. Great job for the `ExpressionResolver` and `ResolverRules`.
I will rebase this PR soon. Thank you for letting me know.

hequn8128 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala:
```diff
@@ -477,6 +484,18 @@ class TableImpl(
+  private[flink] def resolveCalls(fields: Seq[Expression]): Seq[Expression] = {
+    val outputFieldReferences = operationTree.asInstanceOf[LogicalNode]
+      .output.map(a => new UnresolvedFieldReferenceExpression(a.name))
+    val expander = new ColumnsOperationExpander(outputFieldReferences)
```
Due to the refactor of #8062, the `getOutputFieldReferences()` method will not be used by other places, so I merged `getOutputFieldReferences()` and `resolveCalls()` into one.

dawidwys replied:
Why do we need to expands columns here? It seems unnecessary for me. The only reason for the `resolveCalls` in `tableImpl` is because we need to split expressions into `projections`, `aggregates` and `windowProperties`. We don't need to expand columns for that.

hequn8128 replied:
We need to expand columns before `extractFieldReferences`, so that to get the correct field references.

hequn8128 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala:
```diff
@@ -630,18 +672,25 @@ class OverWindowedTableImpl(
 
+  private def expandColumnsInOverWindow(overWindows: Seq[OverWindow]): Seq[OverWindow] = {
+    overWindows.map { e =>
+      val expanedPartitioning = tableImpl.resolveCalls(e.getPartitioning)
```
Remove this method due to #8062

hequn8128 commented:
@sunjincheng121 @dawidwys I have updated the PR. Would be great if you can take another look. :-)

sunjincheng121 commented on docs/dev/table/functions.md:
```diff
@@ -5800,3 +5800,284 @@ For Table API, please use `_` for spaces (e.g., `DAY_TO_HOUR`).
 
 {% top %}
+
+Column Operations
```
Column Functions ?

sunjincheng121 replied:
I am also hesitating, because all of our functions can also be called expression, such as `=`, `<`, `>` These are Expressions, but our documentation is still called Comparison Functions, so I feel Expression is macro, and the functions are relatively specific. What's your opinion?

hequn8128 replied:
Both Functions and Expressions make sense here. The columns() method is an Expression and also a Function. However, if we decided to put this method in `Built-In Functions`, naming it column Functions may be more consistent with the title? @dawidwys What do you think?

hequn8128 replied:
> Could we convert this method to another Visitor? 

@dawidwys Considering your this comment. I think we can rename the `ColumnsOperationExpander` to `ColumnsFunctionExpander` and add a `ColumnsExpressionExpander`?

sunjincheng121 commented on docs/dev/table/functions.md:
```diff
@@ -5800,3 +5800,284 @@ For Table API, please use `_` for spaces (e.g., `DAY_TO_HOUR`).
+</div>
+</div>
+
+The column operations can be used in all places where column fields are expected, such as `select, groupBy, orderBy, UDX etc.` e.g.:
```
UDX -> UDFs ?

sunjincheng121 replied:
`column operations` -> `column functions`

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala:
```diff
@@ -564,26 +575,47 @@ class WindowGroupedTableImpl(
       fields)
   }
 
+  private def expandColumnsInGroupWindow(window: GroupWindow): GroupWindow = {
```
Why do we need it?

hequn8128 replied:
The time filed in GroupWindow can also contains a column operation. For example, 
```
      .window(Slide.over("3.milli").every("10.milli").on("columns(b)").as("w"))
```
And time field are passed to the `extractFieldReferences()`:
```
    val projectFields = extractFieldReferences(
      (expressionsWithResolvedCalls ++ resolvedGroupKeys :+ resolvedWindow.getTimeField))
```
if we don't expand column operations for time field, `extractFieldReferences` would be failed.

dawidwys replied:
I think we should rather modify the `extractFieldReference` to extract also column expressions (They are actually field references). That way we will have a more unified approach and we can use the rule to expand them only in the `ExpressionResolver`

hequn8128 replied:
This would bring some duplicate code. However, it would also make sense. I will try this.

dawidwys replied:
But the code in `extractFieldReference` would be super simple. I haven't checked it, but this should work:

```
		@Override
		public List<Expression> visitCall(CallExpression call) {
			FunctionDefinition functionDefinition = call.getFunctionDefinition();
			if (WINDOW_PROPERTIES.contains(functionDefinition)) {
				return Collections.emptyList();
			} else if (COLUMN_OPERATIONS.contains(functionDefinition)) {
				return call;
			} else {
				return call.getChildren()
					.stream()
					.flatMap(c -> c.accept(this).stream())
					.distinct()
					.collect(Collectors.toList());
			}
		}
```

hequn8128 replied:
We can't return the column expression directly. We need to expand it as all references need to be distincted.
```
	public static List<Expression> extractFieldReferences(List<Expression> expressions) {
		return expressions.stream()
			.flatMap(expr -> expr.accept(referenceExtractor).stream())
			.distinct()
			.collect(Collectors.toList());
	}
```

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/PlannerExpressionConverter.scala:
```diff
@@ -94,6 +94,12 @@ class PlannerExpressionConverter private extends ApiExpressionVisitor[PlannerExp
 
           case AS =>
             assert(args.size >= 2)
+            args.drop(1).foreach {
```
Why do we validate it here? The whole `PlannerExpressionConverter` is already past validationPhase.

hequn8128 replied:
Ok, I validate it because of the test, i.e., `testInvalidRenameColumns()` in `ColumnsOperationValidationTest`. I think you make a good point here, maybe we need to validate the as in Column Expander Rule?

hequn8128 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ColumnsOperationExpander.java:
```diff
@@ -0,0 +1,177 @@
+	/**
+	 * Expand the columns expression in the input Expression.
+	 */
+	private List<UnresolvedReferenceExpression> resolveSingleArg(Expression arg) {
```
Good idea. Let me try this.

twalthr commented:
Thanks for the contribution @hequn8128. I haven't had a look at the code but at the API again.

I still find the `columns(...) vs. -columns(...)` confusing. Why not making it more explicit and fluent, e.g. `withColumns(...) vs. withoutColumns(...)`. There are a couple of advantages:

1. It reads nicely: "select with columns 1 to 5" or "udagg with columns 1 to 5".
2. When I saw the design doc the first time, I thought users can do things like `columns(1 to 5)-columns(2)` (so not just unary minus).
2. It avoids ambiguity. Because a column operation with a single column could be misinterpreted. Take `select(-columns('a))` for a schema `('a, 'b)` as an example. Given that a `columns(...)` is just defined as a replacement operation. This could be interpreted as `select(-'a)` or `select('b)`.
3. We already need to keep a new Java DSL in mind. `withColumns()` and `withoutColumns()` can be easily expressed in Java. The current design would look like `select(minus(columns(...)))`.

What do you think?

sunjincheng121 commented:
At the beginning of the design, we refer to the operation of the R language column, using the `-` flag in R, and the Scala language is also easy to support `-`, of course, we did not consider the upcoming new Java DSL, although `columns(...)` and `-columns(...)` are very concise, but for the description of java DSL, I agree with Timo's proposal.

hequn8128 commented:
@twalthr @sunjincheng121 OK. Let's do it with `withColumns()` and `withoutColumns()`. I will update this PR once [FLINK-12174](https://github.com/apache/flink/pull/8162) is merged.

Best, Hequn

hequn8128 commented:
@sunjincheng121 @twalthr @dawidwys Thanks a lot for your great suggestions. I have updated the PR and rebased to the master. Would be great if you can take another look. Thank you very much!

Best, Hequn

hequn8128 commented:
@sunjincheng121 Thank you for the review. I have renamed these "column operation" to "column function". So in Table API, column operations means Table.addColumns()/Table.dropColumns, and column functions means withColumns/withoutColumns. I think it's more clear now.

Best, Hequn

sunjincheng121 commented:
+1 to merged..

sunjincheng121 commented:
@flinkbot approve all
